bokehjs/src/coffee/models/sources/column_data_source.coffee

```diff
@@ -49,13 +50,58 @@ export stream_to_column = (col, new_col, rollover) ->
   return concat_typed_arrays(col, tmp)
 
 # exported for testing
-export patch_to_column = (col, patch) ->
-  for i in [0...patch.length]
-    [ind, value] = patch[i]
-    col[ind] = value
+export slice = (ind, length) ->
+  if isObject(ind)
+    return [ind.start ? 0, ind.stop ? length, ind.step ? 1]
+  return [start, stop, step] = [ind, ind+1, 1]
 
+# exported for testing
+export patch_to_column = (col, patch, shapes) ->
```

Comment:
E.g. in this example:
```
 +            # replace the entire 10th column of the 2nd array:
 +
 +              +------------- index of item in column data source
 +              |
 +              |     +------- row subindex into array item
 +              |     |
 +              |     |     +- column subindex into array item
 +              V     V     V
 +            ([2, slice(None), 10], new_values)
```
the `index` on the JS side is `[2, {start:null, stop:null, step:null}, 10]`